### ISSUE TITLE
[BUG] Codex Storage node with prover crashes - Unable to get send connection for peer message not sent

### ISSUE THREAD

**Issue by veaceslavdoina** [codex-storage/nim-codex#914]
**Describe the bug**
Codex Storage node crashes periodically.

**To Reproduce**
1. Run Codex Storage node
2. Make storage availability
3. Fill storage requests (?)
4. Node will crash periodically(?)

**Expected behavior**
Codex should not crash.

**Screenshots**
<details>
<summary>screenshots</summary>

### Logs
<img width="1048" alt="Screenshot 2024-09-24 at 22 57 04" src="https://github.com/user-attachments/assets/9d02a307-4fa0-4a81-89c9-eda404d6372f">

### All nodes
<img width="1510" alt="Screenshot 2024-09-24 at 22 52 55" src="https://github.com/user-attachments/assets/03590a17-af15-4bcc-8b15-675e6622c078">

### Node 10
<img width="1508" alt="Screenshot 2024-09-24 at 22 49 23" src="https://github.com/user-attachments/assets/68e728e0-7bf3-4b3c-ae08-26e8b566fbd9">
</details>

**Environment:**
 - OS: `24.04.1 LTS`
 - Version: `0.1.4`
 - Tag/Commit: `484124d`
 - Arch: `amd64`

**Additional context**
We are running 10 Codex Storage nodes and they all crashes periodically.
```shell
codex persistence prover
```

Please see attached full log of one of the crash - [codex-10-1-crash.txt](https://github.com/user-attachments/files/17120201/codex-10-1-crash.txt).

**Comment by benbierens** [comment]
I've just now seen the same issue in my autoclient-node.
Important: This node did not run prover. This issue seems unrelated to prover.
Autoclient-node is run with persistence, and does create purchases. But it does not store and does not create proofs.
I think this one needs to be increased in priority...

**Comment by benbierens** [comment]
I have a sneaky suspicion...
I think the bug might occur _sometimes_, when:
1. a number of hosts are downloading a slot from a client
2. a host completes the download first, submits proof and claims the slot
3. other hosts abort their downloads and drop the connection
I suspect the sudden drop triggers some code in the client that has a race condition, somewhere related to
`networkpeer.nim` line 35 = `proc connected*(b: NetworkPeer): bool`
My guess is that proc connected returns true right _before_ the connection is dropped. The client proceeds as if the connection is alive and runs into a null-ref exception.
I'm throwing tests at this to try to reproduce and gather some evidence.

**Comment by benbierens** [comment]
Test run: All hosts are crashing at the same time, with the same error. The client remains alive.
Tinkering with the logging, then running more tests...

**Comment by benbierens** [comment]
In some cases, client also crashes. In some cases, not all hosts crash.
I have attempted to reproduce the issue without using marketplace by:
 - Adding a debug/crash API call that crashes the node on command. Start a swarm with a large file everyone is downloading, then crash one node with the API call: All other nodes remain alive after the suddenly lost connection.
 - Adding a debug/cancel API call to cancel the pending block and/or stream futures on command. Again: large file, swarm, download, and call cancel on one node. Again all nodes remain alive and functional.
 - Simply interrupting a download stream from the HTTP client. Also no issues.

I checked CancelledException propagation in block exchange code. It may be lacking in some cases, but trying to add it made no difference.
